### PR TITLE
[C++ worker] Remove some api dependencies on ray core

### PR DESCRIPTION
**Pull request by qicosmos**
## Why are these changes needed?

We should support safe c++ worker API according the [doc](https://docs.google.com/document/d/1pBRUm-w2LKXTte8uXmvQ23l7gOJV2IOwirCTPjtPGWk/edit#heading=h.86fui9aqjkh6), this pr remove some dependencies on ray core, such as: 

ErrorType, WorkerType, Language, TaskType, Buffer, LocalMemoryBuffer, RayObject, TaskArg, TaskArgByReference, TaskArgByValue.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests

### CODE REVIEW

File: cpp/include/ray/api/common_types.h [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+#pragma once
+#include <memory>
+#include <string>
+#include "ray/core.h"
+
+namespace ray {
+namespace api {
+
+enum ErrorType : int {
```
Comment (kfstorm) [inline cpp/include/ray/api/common_types.h:23]: why does this need to be included in API?

File: cpp/include/ray/api/common_types.h [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+  WORKER_DIED = 0,
+  ACTOR_DIED = 1,
+  OBJECT_UNRECONSTRUCTABLE = 2,
+  TASK_EXECUTION_EXCEPTION = 3,
+  OBJECT_IN_PLASMA = 4,
+  TASK_CANCELLED = 5,
+  ACTOR_CREATION_FAILED = 6,
+};
+
+enum WorkerType : int {
```
Comment (kfstorm) [inline cpp/include/ray/api/common_types.h:33]: why does this need to be included in API?

File: cpp/include/ray/api/common_types.h [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+  RESTORE_WORKER = 3,
+};
+
+enum Language : int {
+  PYTHON = 0,
+  JAVA = 1,
+  CPP = 2,
+};
+
+enum TaskType : int {
```
Comment (kfstorm) [inline cpp/include/ray/api/common_types.h:46]: why does this need to be included in API?

File: cpp/include/ray/api/common_types.h [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+  NORMAL_TASK = 0,
+  ACTOR_CREATION_TASK = 1,
+  ACTOR_TASK = 2,
+  DRIVER_TASK = 3,
+};
+
+class Buffer {
```
Comment (kfstorm) [inline cpp/include/ray/api/common_types.h:53]: why does this need to be included in API?

File: cpp/include/ray/api/common_types.h [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+  bool operator==(const Buffer &rhs) const {
+    if (this->Size() != rhs.Size()) {
+      return false;
+    }
+
+    return this->Size() == 0 || memcmp(Data(), rhs.Data(), Size()) == 0;
+  }
+};
+
+class LocalMemoryBuffer : public Buffer {
```
Comment (kfstorm) [inline cpp/include/ray/api/common_types.h:77]: why does this need to be included in API?

File: cpp/include/ray/api/common_types.h [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+  uint8_t *data_;
+  /// Size of the buffer.
+  size_t size_;
+  /// Whether this buffer holds a copy of data.
+  bool has_data_copy_;
+  /// This is only valid when `should_copy` is true.
+  uint8_t *buffer_ = NULL;
+};
+
+class RayObject {
```
Comment (kfstorm) [inline cpp/include/ray/api/common_types.h:131]: why does this need to be included in API?

File: cpp/include/ray/api/common_types.h [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+struct Address {
+  std::string raylet_id;
+  std::string ip_address;
+  int32_t port;
+  // Optional unique id for the worker.
+  std::string worker_id;
+};
+
+enum class ArgType { ArgByRef, ArgByValue };
```
Comment (kfstorm) [inline cpp/include/ray/api/common_types.h:183]: why does this need to be included in API?

File: cpp/include/ray/api/common_types.h [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    return size;
+  }
+
+ private:
+  std::shared_ptr<Buffer> data_;
+  std::shared_ptr<Buffer> metadata_;
+  const std::vector<ObjectID> nested_ids_;
+};
+
+struct Address {
```
Comment (kfstorm) [inline cpp/include/ray/api/common_types.h:175]: why does this need to be included in API?

File: cpp/include/ray/api/common_types.h [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+  std::string ip_address;
+  int32_t port;
+  // Optional unique id for the worker.
+  std::string worker_id;
+};
+
+enum class ArgType { ArgByRef, ArgByValue };
+
+/// Argument of a task.
+class TaskArg {
```
Comment (kfstorm) [inline cpp/include/ray/api/common_types.h:186]: why does this need to be included in API?

File: cpp/include/ray/api/common_types.h [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+  virtual const ObjectID GetObjectID() { return {}; }
+  virtual const Address GetAddress() { return {}; }
+  virtual std::shared_ptr<RayObject> GetValue() { return nullptr; }
+
+ protected:
+  ArgType arg_type_;
+};
+
+class TaskArgByReference : public TaskArg {
```
Comment (kfstorm) [inline cpp/include/ray/api/common_types.h:201]: why does this need to be included in API?

File: cpp/include/ray/api/common_types.h [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+  const ObjectID GetObjectID() { return id_; }
+  const Address GetAddress() { return owner_address_; }
+
+ private:
+  /// Id of the argument if passed by reference, otherwise nullptr.
+  const ObjectID id_;
+  const Address owner_address_;
+};
+
+class TaskArgByValue : public TaskArg {
```
Comment (kfstorm) [inline cpp/include/ray/api/common_types.h:220]: why does this need to be included in API?

File: cpp/include/ray/api/common_types.h (hunk 2 above)
Comment (qicosmos) [reply cpp/include/ray/api/common_types.h:33]: api::ray_config need it.

File: cpp/include/ray/api/common_types.h (hunk 3 above)
Comment (qicosmos) [reply cpp/include/ray/api/common_types.h]: No need, removed.

File: cpp/include/ray/api/common_types.h [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    buf = std::move(rhs.buf);
+    id = rhs.id;
+  }
+
+  TaskArg(const TaskArg &) = delete;
+  TaskArg &operator=(TaskArg const &) = delete;
+  TaskArg &operator=(TaskArg &&) = delete;
+
+  boost::optional<msgpack::sbuffer> buf;
+  ObjectID id;
```
Comment (kfstorm) [inline cpp/include/ray/api/common_types.h:52]: Should it be optional as well?

File: cpp/include/ray/api/function_manager.h [hunk 13]
```diff
@@ -49,7 +50,7 @@ inline static msgpack::sbuffer PackError(std::string error_msg) {
   msgpack::sbuffer sbuffer;
   msgpack::packer<msgpack::sbuffer> packer(sbuffer);
   packer.pack(msgpack::type::nil_t());
-  packer.pack(std::make_tuple((int)ray::rpc::ErrorType::TASK_EXECUTION_EXCEPTION,
+  packer.pack(std::make_tuple((int)ray::api::ErrorType::TASK_EXECUTION_EXCEPTION,
```
Comment (kfstorm) [inline cpp/include/ray/api/function_manager.h:53]: Do we have to expose `ErrorType` here? It seems that it's cast to an `int` anyway.

Comment (qicosmos) [reply cpp/include/ray/api/function_manager.h:53]: Do you mean use an integer value instead?

File: cpp/include/ray/api/common_types.h (hunk 12 above)
Comment (qicosmos) [reply cpp/include/ray/api/common_types.h:52]: The optional buf is enough, if buf init means value. else means reference.

Comment (kfstorm) [reply cpp/include/ray/api/common_types.h:52]: Then please add some comments on these fields.

File: cpp/include/ray/api/function_manager.h (hunk 13 above)
Comment (kfstorm) [reply cpp/include/ray/api/function_manager.h:53]: Actually, I don't understand the code. It seems that all errors are converted to `TASK_EXECUTION_EXCEPTION`. Why?

Comment (qicosmos) [reply cpp/include/ray/api/function_manager.h:53]: Because function_namnager has template classes, so the ErrorType can't be hidden in runtime

**Review by kfstorm — APPROVED** [review]

**Comment by kfstorm** [comment]
Windows CI failed but should be unrelated.

//python/ray/tests:test_async                                             FLAKY, failed in 1 out of 2 in 300.1s
//python/ray/tests:test_runtime_env                                      FAILED in 3 out of 3 in 790.5s

**Pull request merged**